packages/mui-material/src/Tabs/Tabs.test.js

```diff
@@ -1397,4 +1334,76 @@ describe('<Tabs />', () => {
       ]);
     });
   });
+
+  describe('dynamic tabs', () => {
+    const pause = (timeout) =>
+      new Promise((resolve) => {
+        setTimeout(() => {
+          resolve();
+        }, timeout);
+      });
+
+    // https://github.com/mui/material-ui/issues/31936
+    it('should not show scroll buttons if a tab added or removed in vertical mode', async function test() {
+      if (isJSDOM) {
+        this.skip();
+      }
+      function DynamicTabs() {
+        const [value, setValue] = React.useState(0);
+        const handleChange = (event, newValue) => {
+          setValue(newValue);
+        };
+        const [tabs, setTabs] = React.useState(['item1', 'item2']);
+        return (
+          <React.Fragment>
+            <button
+              data-testid="add"
+              onClick={() => {
+                setTabs([...tabs, `item${tabs.length + 1}`]);
+              }}
+            >
+              add
+            </button>
+            <button
+              data-testid="delete"
+              onClick={() => {
+                setTabs(tabs.slice(0, tabs.length - 1));
+                setValue(0);
+              }}
+            >
+              delete
+            </button>
+            <Tabs
+              onChange={handleChange}
+              value={value}
+              orientation="vertical"
+              variant="scrollable"
+              scrollButtons
+              style={{ width: '260px' }}
+            >
+              {tabs.map((label, index) => (
+                <Tab key={`tab${index}`} label={label} />
+              ))}
+            </Tabs>
+          </React.Fragment>
+        );
+      }
+      const { container, getByTestId, getAllByRole } = render(<DynamicTabs />);
+      const addButton = getByTestId('add');
+      const deleteButton = getByTestId('delete');
+
+      fireEvent.click(addButton);
+      expect(hasLeftScrollButton(container)).to.equal(false);
+      expect(hasRightScrollButton(container)).to.equal(false);
+
+      const tabs = getAllByRole('tab');
+      const lastTab = tabs[tabs.length - 1];
+      fireEvent.click(lastTab);
+      await pause(400);
```

Comment:
We never use the real-time in these tests, to **remove**

```suggestion

```

Reply:
I couldn't make the test cover that bug without this, but I'm aware of it. Do you have any alternative solutions by faking the time?

Reply:
Same i tried requestAnimationFrame but it doesn't work so i did the pause(400)

Reply:
I had a look, my process was to: 1. revert all the test changes 2. focus on one failing test, 3. make it pass. Here is what I come up with:

https://github.com/oliviertassinari/material-ui/commit/76209fdcfb4fd5e030821fa9cfd8dc2f92edb797#diff-1d60104297e2e5decebdb9cef7bdbf776affd72d75fa7f263d2d5e387358499aR673

We should be able to fix all the other tests like this. I removed the time mocking in this instance, as it wasn't relevant to the test (what we need to depend on is not time but rafs)

Could we fix the rest? Thanks!

Reply:
> what we need to depend on is not time but rafs

Is it because `waitFor` should be used only in JSDom, and `requestAnimationFrame` is used in the browser environment?

Reply:
> > what we need to depend on is not time but rafs
> 
> Is it because `waitFor` should be used only in JSDom, and `requestAnimationFrame` is used in the browser environment?

Other tests i used the raf and it worked, but for this specific unit test i don't know why we must delay the delete click.

Reply:
@oliviertassinari @ZeeshanTamboli 
Created a PR for this [#38208](https://github.com/mui/material-ui/pull/38208).

I revert to the old implementation of unit tests then I refacto + added the last unit that fails in old implementation of Tabs before IntersectionObserver.